src/filesystem-cache.ts

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2020 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ not
+ * use this file except in compliance with the License. You may obtain a copy
+ of
+ * the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ under
+ * the License.
+ */
+
+'use strict';
+
+import * as fs from 'fs';
+import * as path from 'path';
+import * as Koa from 'koa';
+import {Config} from './config';
+
+type CacheContent = {
+  saved: Date,
+  expires: Date,
+  response: string,
+  payload: string,
+};
+
+function hashCode(s: string): string {
+  let hash = 0;
+  if (s.length === 0) return hash.toString();
+
+  for (let i = 0; i < s.length; i++) {
+    const char = s.charCodeAt(i);
+    hash = ((hash << 5) - hash) + char;
+    hash &= hash;
+  }
+
+  return hash.toString();
+}
+
+export class FilesystemCache {
+  private config: Config;
+  private cacheConfig: { [key: string]: string };
+
+  constructor(config: Config) {
+    this.config = config;
+    this.cacheConfig = this.config.cacheConfig;
+
+    if (!fs.existsSync(this.getDir(''))) {
+      fs.mkdirSync(this.getDir(''));
+    }
+  }
+
+  getDir = (key: string) => {
+    const dir = this.cacheConfig.snapshotDir;
+
+    if (key) {
+      return path.join(dir, key);
+    }
+
+    return dir;
+  }
+
+  async clearCache(key: string) {
+    fs.rmdirSync(this.getDir(key), {recursive: true});
+  }
+
+  async clearAllCache() {
+    fs.rmdirSync(this.getDir(''), {recursive: true});
+  }
+
+  cacheContent(key: string, ctx: Koa.Context) {
+    const response = ctx.response;
```

Comment:
To make naming a bit more clear, what do you think about naming this variable `responseHeaders` or `responseMeta` or something similar? Then the `payload` could either stay with that name or be renamed `responseBody` to make the distinction clearer.